src/lib.rs

```diff
@@ -0,0 +1,528 @@
+        match monitor(router, abort_handles, epsilon).await {
+            Reconnect::NewUpstream(new_pool_addr) => {
+                ProxyState::update_proxy_state_up();
+                pool_addr = Some(new_pool_addr);
+                continue;
+            }
+            Reconnect::NoUpstream => {
+                ProxyState::update_proxy_state_up();
+                pool_addr = None;
+                continue;
+            }
+        };
+    }
+}
+
+async fn monitor(
+    router: &mut Router,
+    abort_handles: Vec<(AbortOnDrop, std::string::String)>,
+    epsilon: Duration,
+) -> Reconnect {
+    let mut should_check_upstreams_latency = 0;
+    loop {
+        if Configuration::monitor() {
+            // Check if a better upstream exist every 100 seconds
+            if should_check_upstreams_latency == 10 * 100 {
+                should_check_upstreams_latency = 0;
+                if let Some(new_upstream) = router.monitor_upstream(epsilon).await {
+                    info!("Faster upstream detected. Reinitializing proxy...");
+                    drop(abort_handles);
+
+                    // Needs a little to time to drop
+                    tokio::time::sleep(tokio::time::Duration::from_millis(10)).await;
+                    return Reconnect::NewUpstream(new_upstream);
+                }
+            }
+            should_check_upstreams_latency += 1;
+        }
+
+        // Monitor finished tasks
+        if let Some((_handle, name)) = abort_handles
+            .iter()
+            .find(|(handle, _name)| handle.is_finished())
+        {
+            error!("Task {:?} finished, Closing connection", name);
+            for (handle, _name) in abort_handles {
+                drop(handle);
+            }
+
+            // Check if the proxy state is down, and if so, reinitialize the proxy.
+            let is_proxy_down = ProxyState::is_proxy_down();
+            if is_proxy_down.0 {
+                error!(
+                    "Status: {:?}. Reinitializing proxy...",
+                    is_proxy_down.1.unwrap_or("Proxy".to_string())
+                );
+                return Reconnect::NoUpstream;
+            } else {
+                return Reconnect::NoUpstream;
+            }
+        }
+
+        // Check if the proxy state is down, and if so, reinitialize the proxy.
+        let is_proxy_down = ProxyState::is_proxy_down();
+        if is_proxy_down.0 {
+            error!(
+                "{:?} is DOWN. Reinitializing proxy...",
+                is_proxy_down.1.unwrap_or("Proxy".to_string())
+            );
+            drop(abort_handles); // Drop all abort handles
+            tokio::time::sleep(tokio::time::Duration::from_millis(10)).await; // Needs a little to time to drop
+            return Reconnect::NoUpstream;
+        }
+
+        tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
+    }
+}
+
+fn check_update_proxy() {
```

Comment:
can u please move all the auto update code into a separate file `auto_update.rs`